lib/improver/wxcode/weather_symbols.py

```diff
@@ -0,0 +1,403 @@
+                e.g. {<node_name>: [<succeed_name>, <fail_name>]}
+            start (string):
+                The node name of the tree root (currently always
+                significant_precipitation).
+            end (int):
+                The weather symbol code to which we are tracing all routes.
+
+        Returns:
+            routes (list):
+                A list of node names that defines the route from the tree root
+                to the weather symbol leaf (end of chain).
+
+        References:
+            Method based upon Python Patterns - Implementing Graphs essay
+            https://www.python.org/doc/essays/graphs/
+        """
+        if route is None:
+            route = []
+
+        route = route + [start]
+        if start == end:
+            return [route]
+        if start not in graph.keys():
+            return []
+
+        routes = []
+        for node in graph[start]:
+            if node not in route:
+                newroutes = WeatherSymbols.find_all_routes(graph, node, end,
+                                                           route)
+                for newroute in newroutes:
+                    routes.append(newroute)
+        return routes
+
+    @staticmethod
+    def create_symbol_cube(cube):
+        """
+        Create an empty weather_symbol cube initialised with -1 across the
+        grid.
+
+        Args:
+            cube (iris.cube.Cube):
+                An x-y slice of one of the input cubes, used to define the
+                size of the weather symbol grid.
+        Returns:
+            symbols (iris.cube.Cube):
+                A cube full of -1 values, with suitable metadata to describe
+                the weather symbols that will fill it.
+        """
+        cube_format = next(cube.slices_over(['threshold']))
+        symbols = cube_format.copy(data=np.full(cube_format.data.shape, -1,
+                                                dtype=np.int))
+
+        symbols.remove_coord('threshold')
+        symbols.attributes.pop('relative_to_threshold')
+        symbols = add_wxcode_metadata(symbols)
+
+        return symbols
+
+    def process(self, cubes):
+        """Apply the decision tree to the input cubes to produce weather
+        symbol output.
+
+        Args:
+            cubes (iris.cube.CubeList):
+                A cubelist containing the diagnostics required for the
+                weather symbols decision tree, these at conincident times.
+
+        Returns:
+            symbols (iris.cube.Cube):
+                A cube of weather symbols.
+        """
+        # Check input cubes contain required data
+        self.check_input_cubes(cubes)
+
+        # Construct graph nodes dictionary
+        graph = {key: [self.queries[key]['succeed'], self.queries[key]['fail']]
+                 for key in self.queries.keys()}
+
+        # Search through tree for all leaves (weather code end points)
+        defined_symbols = []
```

Comment:
I think that these for loops could be replaced by:
```python
from improver.wxcode.wxcode_utilities import WX_DICT
...
defined_symbols = WX_DICT.keys()
```
I realise that we don't currently support all of the weather symbols but I think this is more future-proof, whilst the current method relies upon only the weather symbols being integers.

Reply:
Since the full list contains night as well as day symbols and dust I think it would make more sense to leave this as it is.